# Heroes of Pymoli Data Analysis

For this project, I'll be taking a dataset and analyzing the information to see what insights I can glean from it. This is a dataset called purchase_data.csv, which is data from a game called Heros of Pymoli. This data contains information about in-app purchases users may have made when they were playing the game. 

The first thing I'll do is import the data I'll be working with and take a quick peek to see what it looks like. 

In [1]:

# These are the modules I'll be importing to help my perform my analyis
import pandas as pd
import os

# The dataset is in a file, and I'm loading the path to that file into a variable so that it can be called into my environment for analysis
file_to_load = os.path.join("Resources", "purchase_data.csv")

# Read Purchasing File and store into Pandas data frame
purchases_df = pd.read_csv(file_to_load)



In [2]:
#take a look at the data after it is read to a csv
purchases_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [3]:
#take a look at columns to see if I have any columns with missing data
purchases_df.count()

Purchase ID    780
SN             780
Age            780
Gender         780
Item ID        780
Item Name      780
Price          780
dtype: int64

The data looks good, so I'll start my analysis.

## Player Count

My first task will be to display the total number of players in the dataset. Players may show up multiple times, so I'll provide the unique player count. 


In [4]:
#I'll use the SN (screen name?) column to determine the unique number of players. Screen names may show up multiple times, so I'll get the count of unique, instead of just the count of items in the column. 
player_count = len(purchases_df["SN"].unique())


In [5]:
#Now that I have my count, I'll place the value in a dataframe for neat presentation
player_cnt_summary = pd.DataFrame({"Total Players":[player_count]})
player_cnt_summary

Total Players
0            576

## Purchasing Analysis (Total)

### Calculations

In this section, I'll run some calculations on the dataset and then present it a dataframe. The calculations I'll include are:
* Unique Items
* Average Prices
* Total Purchase 
* Total Revenue

In [6]:
#I'm now going to take a look at the data to obtain some basic summary statistics

#number of unique item IDs.  I initially wanted to look at the item names, but it looks like some items are showing up more than once with different item IDs.  I'll use item IDs instead.
unique_items = len(purchases_df["Item ID"].unique())
unique_items



183

In [7]:
#I'll get the average price of the items in the list
average_price = purchases_df["Price"].mean()
average_price

3.050987179487176

In [8]:
#I'll get the total number of purchases
total_purchases = purchases_df["Purchase ID"].count()
total_purchases

780

In [9]:
#I'll get the total revenue from the purchases
total_revenue  = purchases_df["Price"].sum()
total_revenue

2379.77

### Summary

In this section, I'll take my calculations and place it a dataframe, format it and present it. 

In [10]:
#I'll place all of these into a dataframe for neat presentation
purchase_analysis_df = pd.DataFrame({"Number of Unique Items":[unique_items],
                                    "Average Price": [average_price],
                                    "Number of Purchases": [total_purchases],
                                    "Total Revenue":[total_revenue]})
purchase_analysis_df

Number of Unique Items  Average Price  Number of Purchases  Total Revenue
0                     183       3.050987                  780        2379.77

In [11]:
#next, I'll format my data that should be in currency format to show that way
purchase_analysis_df["Average Price"] = purchase_analysis_df["Average Price"].astype(float).map("${:,.2f}".format)
purchase_analysis_df["Total Revenue"] = purchase_analysis_df["Total Revenue"].astype(float).map("${:,.2f}".format)
purchase_analysis_df


Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     183         $3.05                  780     $2,379.77

## Gender Demographics

### Calculations

For this section, I want to take my data and provide some statistics based on gender information that is available in the dataset.  I'll be looking to calculate the following items
* Percentage and Count of Male Players
* Percentage and Count of Female Players
* Percentage and Count of Other / Non-Disclosed




In [12]:
#I'll take my initial dataframe and group by the gender column 
gender_demo_grpby = purchases_df.groupby(['Gender'])
gender_demo_grpby.count()

Purchase ID   SN  Age  Item ID  Item Name  Price
Gender                                                                 
Female                         113  113  113      113        113    113
Male                           652  652  652      652        652    652
Other / Non-Disclosed           15   15   15       15         15     15

In [13]:
#Here, I'll take my grouped data and get a unique count per gender category
unique_count = gender_demo_grpby["SN"].nunique()
unique_count

Gender
Female                    81
Male                     484
Other / Non-Disclosed     11
Name: SN, dtype: int64

In [14]:
#Next, I'll take my unique counts per gender and divide by the total number of unique users
percentage = unique_count/player_count * 100
percentage


Gender
Female                   14.062500
Male                     84.027778
Other / Non-Disclosed     1.909722
Name: SN, dtype: float64

### Summary

I'll place the percentages I found in a dataframe, format it and present it.

In [15]:
#now, I'll put my results into a dataframe for neat presentation. 
gender_demo_summary = pd.DataFrame({"Total Count":unique_count,
                                 "Percentage of Players": percentage})

gender_demo_summary

Total Count  Percentage of Players
Gender                                                   
Female                          81              14.062500
Male                           484              84.027778
Other / Non-Disclosed           11               1.909722

In [16]:
#now that I have the data, I want to sort it in descending order and format the percentages
gender_demo_sort = gender_demo_summary.sort_values("Total Count", ascending = False)
gender_demo_sort["Percentage of Players"] = gender_demo_sort["Percentage of Players"].astype(float).map("{:.2f}%".format)
gender_demo_sort.head()

Total Count Percentage of Players
Gender                                                  
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%

## Purchasing Analysis (Gender)

### Calculations

In this section, I'll run some calculations on the dataset, grouped by gender, and then present it a dataframe. The calculations I'll include are:
* Unique Items
* Average Prices
* Total Purchase 
* Total Revenue

In [17]:
#I'm going to take a look at my grouped by data again.
gender_demo_grpby.count()

Purchase ID   SN  Age  Item ID  Item Name  Price
Gender                                                                 
Female                         113  113  113      113        113    113
Male                           652  652  652      652        652    652
Other / Non-Disclosed           15   15   15       15         15     15

In [18]:
#I'm going to put the count of purchases in a variable
gender_purchase_count = gender_demo_grpby["Purchase ID"].count()
gender_purchase_count

Gender
Female                   113
Male                     652
Other / Non-Disclosed     15
Name: Purchase ID, dtype: int64

In [19]:
#I'm going to determing the average purchase price per gender
gender_avg_price = gender_demo_grpby["Price"].mean()
gender_avg_price

Gender
Female                   3.203009
Male                     3.017853
Other / Non-Disclosed    3.346000
Name: Price, dtype: float64

In [20]:
#I'm going to determine the total purchase value
gender_total_price = gender_demo_grpby["Price"].sum()
gender_total_price

Gender
Female                    361.94
Male                     1967.64
Other / Non-Disclosed      50.19
Name: Price, dtype: float64

In [21]:
#I'm going to determine the purchase total per person, by gender
avg_total_price_person = gender_total_price/unique_count
avg_total_price_person

Gender
Female                   4.468395
Male                     4.065372
Other / Non-Disclosed    4.562727
dtype: float64

### Summary

In this section, I'll take my calculations and place it a dataframe, format it and present it.

In [22]:
#I'll put these values into a dataframe
gender_purchase_summary = pd.DataFrame({"Purchase Count": gender_purchase_count,
                                 "Average Purchase Price": gender_avg_price,
                                 "Total Purchase Value": gender_total_price,
                                 "Avg Total Purchase per Person": avg_total_price_person})

gender_purchase_summary.head()

Purchase Count  Average Purchase Price  \
Gender                                                          
Female                            113                3.203009   
Male                              652                3.017853   
Other / Non-Disclosed              15                3.346000   

                       Total Purchase Value  Avg Total Purchase per Person  
Gender                                                                      
Female                               361.94                       4.468395  
Male                                1967.64                       4.065372  
Other / Non-Disclosed                 50.19                       4.562727

In [23]:
#I'll format some of the numbers in my dataframe to present like currency.
gender_purchase_summary["Average Purchase Price"] = gender_purchase_summary["Average Purchase Price"].astype(float).map("${:.2f}".format)
gender_purchase_summary["Total Purchase Value"] = gender_purchase_summary["Total Purchase Value"].astype(float).map("${:.2f}".format)
gender_purchase_summary["Avg Total Purchase per Person"] = gender_purchase_summary["Avg Total Purchase per Person"].astype(float).map("${:.2f}".format)
gender_purchase_summary

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                              $1967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

### Calculations

Next, I'll take a look at my data based on age groupings. In order to do that, I'll complete the following steps
* Establish bins for ages
* Categorize the existing players using the age bins.
* Calculate the numbers and percentages by age group

In [24]:
#I'll go back to my original dataframe to create some bins based on the ages of the players
purchases_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [25]:
#there are some users that have made multiple purchases, so I'll create a new dataframe that contains an entry per user and their age
purchase_data_ages = purchases_df.loc[:,["SN", "Age"]]
purchase_data_ages

SN  Age
0          Lisim78   20
1      Lisovynya38   40
2       Ithergue48   24
3    Chamassasya86   24
4        Iskosia90   23
..             ...  ...
775     Aethedru70   21
776         Iral74   21
777     Yathecal72   20
778        Sisur91    7
779      Ennrian78   24

[780 rows x 2 columns]

In [26]:
#There are some users that have made multiple purchases, and will show up in the data more than onece, so I'll get the unique list of users and thier ages by dropping the duplicates
purchase_data_ages_unique = purchase_data_ages.drop_duplicates()
purchase_data_ages_unique

SN  Age
0          Lisim78   20
1      Lisovynya38   40
2       Ithergue48   24
3    Chamassasya86   24
4        Iskosia90   23
..             ...  ...
773         Hala31   21
774     Jiskjask80   11
775     Aethedru70   21
777     Yathecal72   20
778        Sisur91    7

[576 rows x 2 columns]

In [27]:
#I want to figure out the max age
print(purchase_data_ages_unique["Age"].max())

45


In [28]:
# The max age is, 45, so I'll I want to create the following bins
bins = [0, 9, 14, 19, 24, 29, 34, 39, 50 ]

In [29]:
#I'll create some labels for my bins so that I can use the values in a field in my dataframe
group_names = ["<10","10-14","15-19", "20-24", "25-29","30-34","35-39","40+"]

In [30]:
#I'll use the cut method to create the bins. Each of the values in my data will have a bin/grouping associated with it, and that bin will be placed in to a column for analysis.
purchase_data_ages_unique["Age Group"] = pd.cut(purchase_data_ages_unique["Age"], bins, labels=group_names)
purchase_data_ages_unique


C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


SN  Age Age Group
0          Lisim78   20     20-24
1      Lisovynya38   40       40+
2       Ithergue48   24     20-24
3    Chamassasya86   24     20-24
4        Iskosia90   23     20-24
..             ...  ...       ...
773         Hala31   21     20-24
774     Jiskjask80   11     10-14
775     Aethedru70   21     20-24
777     Yathecal72   20     20-24
778        Sisur91    7       <10

[576 rows x 3 columns]

In [31]:
#next, I'll groupby my bins
age_groupby = purchase_data_ages_unique.groupby("Age Group").count()
age_groupby

SN  Age
Age Group          
<10         17   17
10-14       22   22
15-19      107  107
20-24      258  258
25-29       77   77
30-34       52   52
35-39       31   31
40+         12   12

In [32]:
#I'm going to put the age values into a variable
age_groupby_age = age_groupby["Age"]
age_groupby_age

Age Group
<10       17
10-14     22
15-19    107
20-24    258
25-29     77
30-34     52
35-39     31
40+       12
Name: Age, dtype: int64

In [33]:
#next, I'll find the percentage of unique players by my bins
perc_of_play = age_groupby["SN"] /player_count * 100
perc_of_play

Age Group
<10       2.951389
10-14     3.819444
15-19    18.576389
20-24    44.791667
25-29    13.368056
30-34     9.027778
35-39     5.381944
40+       2.083333
Name: SN, dtype: float64

### Summary

In this section, I'll take my calculations and place it a dataframe, format it and present it.

In [34]:
#Now that I have the data, I'll create a dataframe for neat presentation
age_purchase_summary = pd.DataFrame({"Total Count": age_groupby_age,
                                 "Percentage of Players": perc_of_play})
age_purchase_summary

Total Count  Percentage of Players
Age Group                                    
<10                 17               2.951389
10-14               22               3.819444
15-19              107              18.576389
20-24              258              44.791667
25-29               77              13.368056
30-34               52               9.027778
35-39               31               5.381944
40+                 12               2.083333

In [35]:
#I'll format some of the numbers in my dataframe to present like percentages.
age_purchase_summary["Percentage of Players"] = age_purchase_summary["Percentage of Players"].astype(float).map("{:.2f}%".format)
age_purchase_summary

Total Count Percentage of Players
Age Group                                   
<10                 17                 2.95%
10-14               22                 3.82%
15-19              107                18.58%
20-24              258                44.79%
25-29               77                13.37%
30-34               52                 9.03%
35-39               31                 5.38%
40+                 12                 2.08%

## Purchasing Analysis (Age)

### Calculations

Next, I'll take a look at my data based on age groupings. In order to do that, I'll complete the following steps

* Bin the purchase_data data frame by age
* Run basic calculations to obtain
    * Purchase Count
    * Avg Purchase Price
    * Total Purchase Price
    * Avg Total Purchase per Person

In [36]:
#I'll use my original dataframe to create some bins and group by those bins.  I'll use the same bin parameters
# The max age is, 45, so I'll I want to create the following bins
bins = [0, 9, 14, 19, 24, 29, 34, 39, 50 ]
#I'll create some labels for my bins so that I can use the values in a field in my dataframe
group_names = ["<10","10-14","15-19", "20-24", "25-29","30-34","35-39","40+"]

In [37]:
#I'll use the cut method to create the bins. Each of the values in my data will have a bin/grouping associated with it, and that bin will be placed in to a column for analysis.
purchases_df["Age Group"] = pd.cut(purchases_df["Age"], bins, labels=group_names)
purchases_df

Purchase ID             SN  Age  Gender  Item ID  \
0              0        Lisim78   20    Male      108   
1              1    Lisovynya38   40    Male      143   
2              2     Ithergue48   24    Male       92   
3              3  Chamassasya86   24    Male      100   
4              4      Iskosia90   23    Male      131   
..           ...            ...  ...     ...      ...   
775          775     Aethedru70   21  Female       60   
776          776         Iral74   21    Male      164   
777          777     Yathecal72   20    Male       67   
778          778        Sisur91    7    Male      101   
779          779      Ennrian78   24    Male       50   

                                     Item Name  Price Age Group  
0    Extraction, Quickblade Of Trembling Hands   3.53     20-24  
1                            Frenzied Scimitar   1.56       40+  
2                                 Final Critic   4.88     20-24  
3                                  Blindscythe   3.27     20-24  
4                                         Fury   1.44     20-24  
..                                         ...    ...       ...  
775                                       Wolf   3.54     20-24  
776                           Exiled Doomblade   1.63     20-24  
777      Celeste, Incarnation of the Corrupted   3.46     20-24  
778                               Final Critic   4.19       <10  
779                                       Dawn   4.60     20-24  

[780 rows x 8 columns]

In [38]:
#Next, I'll group by my new column, Age Group
purchases_age_groupby = purchases_df.groupby("Age Group")
purchases_age_groupby.count()

Purchase ID   SN  Age  Gender  Item ID  Item Name  Price
Age Group                                                          
<10                 23   23   23      23       23         23     23
10-14               28   28   28      28       28         28     28
15-19              136  136  136     136      136        136    136
20-24              365  365  365     365      365        365    365
25-29              101  101  101     101      101        101    101
30-34               73   73   73      73       73         73     73
35-39               41   41   41      41       41         41     41
40+                 13   13   13      13       13         13     13

In [39]:
#I'm going to put the count of purchases, per age grouping, in a variable
age_purchase_count = purchases_age_groupby["Purchase ID"].count()
age_purchase_count

Age Group
<10       23
10-14     28
15-19    136
20-24    365
25-29    101
30-34     73
35-39     41
40+       13
Name: Purchase ID, dtype: int64

In [40]:
#I'm going to determing the average purchase price per age grouping.
age_avg_price = purchases_age_groupby["Price"].mean()
age_avg_price

Age Group
<10      3.353478
10-14    2.956429
15-19    3.035956
20-24    3.052219
25-29    2.900990
30-34    2.931507
35-39    3.601707
40+      2.941538
Name: Price, dtype: float64

In [41]:
#I'm going to determine the total purchase value, per age grouping.
age_total_price = purchases_age_groupby["Price"].sum()
age_total_price

Age Group
<10        77.13
10-14      82.78
15-19     412.89
20-24    1114.06
25-29     293.00
30-34     214.00
35-39     147.67
40+        38.24
Name: Price, dtype: float64

In [42]:
#I'm going to determine the purchase total per person, per age grouping. I have to use the unique number of users per grouping bin for this math
avg_total_price_person_age = age_total_price/age_groupby_age
avg_total_price_person_age

Age Group
<10      4.537059
10-14    3.762727
15-19    3.858785
20-24    4.318062
25-29    3.805195
30-34    4.115385
35-39    4.763548
40+      3.186667
dtype: float64

### Summary

In this section, I'll take my calculations and place it a dataframe, format it and present it.

In [43]:
#I'll put these values in a dataframe for presentation
age_group_purchase_summary = pd.DataFrame({"Purchase Count": age_purchase_count,
                                 "Average Purchase Price": age_avg_price,
                                 "Total Purchase Value": age_total_price,
                                 "Avg Total Purchase per Person": avg_total_price_person_age})

age_group_purchase_summary

Purchase Count  Average Purchase Price  Total Purchase Value  \
Age Group                                                                 
<10                    23                3.353478                 77.13   
10-14                  28                2.956429                 82.78   
15-19                 136                3.035956                412.89   
20-24                 365                3.052219               1114.06   
25-29                 101                2.900990                293.00   
30-34                  73                2.931507                214.00   
35-39                  41                3.601707                147.67   
40+                    13                2.941538                 38.24   

           Avg Total Purchase per Person  
Age Group                                 
<10                             4.537059  
10-14                           3.762727  
15-19                           3.858785  
20-24                           4.318062  
25-29                           3.805195  
30-34                           4.115385  
35-39                           4.763548  
40+                             3.186667

In [44]:
#I'll format some of the numbers in my dataframe to present like currency.
age_group_purchase_summary["Average Purchase Price"] = age_group_purchase_summary["Average Purchase Price"].astype(float).map("${:.2f}".format)
age_group_purchase_summary["Total Purchase Value"] = age_group_purchase_summary["Total Purchase Value"].astype(float).map("${:.2f}".format)
age_group_purchase_summary["Avg Total Purchase per Person"] = age_group_purchase_summary["Avg Total Purchase per Person"].astype(float).map("${:.2f}".format)
age_group_purchase_summary

Purchase Count Average Purchase Price Total Purchase Value  \
Age Group                                                               
<10                    23                  $3.35               $77.13   
10-14                  28                  $2.96               $82.78   
15-19                 136                  $3.04              $412.89   
20-24                 365                  $3.05             $1114.06   
25-29                 101                  $2.90              $293.00   
30-34                  73                  $2.93              $214.00   
35-39                  41                  $3.60              $147.67   
40+                    13                  $2.94               $38.24   

          Avg Total Purchase per Person  
Age Group                                
<10                               $4.54  
10-14                             $3.76  
15-19                             $3.86  
20-24                             $4.32  
25-29                             $3.81  
30-34                             $4.12  
35-39                             $4.76  
40+                               $3.19

## Top Spenders

### Calculations

In this section, I'll Run basic calculations to obtain the results and present the top 5 items, by total purchase value.  I'll be presenting:

* Purchase Count
* Average Purchase Price
* Total Purchase Value

In [45]:
#Take a look at my main dataframe
purchases_df

Purchase ID             SN  Age  Gender  Item ID  \
0              0        Lisim78   20    Male      108   
1              1    Lisovynya38   40    Male      143   
2              2     Ithergue48   24    Male       92   
3              3  Chamassasya86   24    Male      100   
4              4      Iskosia90   23    Male      131   
..           ...            ...  ...     ...      ...   
775          775     Aethedru70   21  Female       60   
776          776         Iral74   21    Male      164   
777          777     Yathecal72   20    Male       67   
778          778        Sisur91    7    Male      101   
779          779      Ennrian78   24    Male       50   

                                     Item Name  Price Age Group  
0    Extraction, Quickblade Of Trembling Hands   3.53     20-24  
1                            Frenzied Scimitar   1.56       40+  
2                                 Final Critic   4.88     20-24  
3                                  Blindscythe   3.27     20-24  
4                                         Fury   1.44     20-24  
..                                         ...    ...       ...  
775                                       Wolf   3.54     20-24  
776                           Exiled Doomblade   1.63     20-24  
777      Celeste, Incarnation of the Corrupted   3.46     20-24  
778                               Final Critic   4.19       <10  
779                                       Dawn   4.60     20-24  

[780 rows x 8 columns]

In [46]:
#Group by SN to see information per user
purchases_sn_groupby = purchases_df.groupby("SN")
purchases_sn_groupby.count()

Purchase ID  Age  Gender  Item ID  Item Name  Price  Age Group
SN                                                                           
Adairialis76             1    1       1        1          1      1          1
Adastirin33              1    1       1        1          1      1          1
Aeda94                   1    1       1        1          1      1          1
Aela59                   1    1       1        1          1      1          1
Aelaria33                1    1       1        1          1      1          1
...                    ...  ...     ...      ...        ...    ...        ...
Yathecal82               3    3       3        3          3      3          3
Yathedeu43               2    2       2        2          2      2          2
Yoishirrala98            1    1       1        1          1      1          1
Zhisrisu83               2    2       2        2          2      2          2
Zontibe81                3    3       3        3          3      3          3

[576 rows x 7 columns]

In [47]:
#put the purchase ID counts in a variable
top_group_purch_count = purchases_sn_groupby["Purchase ID"].count()
top_group_purch_count

SN
Adairialis76     1
Adastirin33      1
Aeda94           1
Aela59           1
Aelaria33        1
                ..
Yathecal82       3
Yathedeu43       2
Yoishirrala98    1
Zhisrisu83       2
Zontibe81        3
Name: Purchase ID, Length: 576, dtype: int64

In [48]:
#put the avgerage prices in a variable, per SN
top_group_avg = purchases_sn_groupby["Price"].mean()
top_group_avg

SN
Adairialis76     2.280000
Adastirin33      4.480000
Aeda94           4.910000
Aela59           4.320000
Aelaria33        1.790000
                   ...   
Yathecal82       2.073333
Yathedeu43       3.010000
Yoishirrala98    4.580000
Zhisrisu83       3.945000
Zontibe81        2.676667
Name: Price, Length: 576, dtype: float64

In [49]:
#put the avgerage prices in a variable, per SN
top_group_total = purchases_sn_groupby["Price"].sum()
top_group_total

SN
Adairialis76     2.28
Adastirin33      4.48
Aeda94           4.91
Aela59           4.32
Aelaria33        1.79
                 ... 
Yathecal82       6.22
Yathedeu43       6.02
Yoishirrala98    4.58
Zhisrisu83       7.89
Zontibe81        8.03
Name: Price, Length: 576, dtype: float64

### Summary

In this section, I'll take my calculations and place it a dataframe, format it and present only the top 5 items, based on total purchase value

In [50]:
#I'll put these values in a dataframe for presentation
top_purchase_summary = pd.DataFrame({"Purchase Count": top_group_purch_count,
                                 "Average Purchase Price": top_group_avg,
                                 "Total Purchase Value": top_group_total})

top_purchase_summary

Purchase Count  Average Purchase Price  Total Purchase Value
SN                                                                         
Adairialis76                1                2.280000                  2.28
Adastirin33                 1                4.480000                  4.48
Aeda94                      1                4.910000                  4.91
Aela59                      1                4.320000                  4.32
Aelaria33                   1                1.790000                  1.79
...                       ...                     ...                   ...
Yathecal82                  3                2.073333                  6.22
Yathedeu43                  2                3.010000                  6.02
Yoishirrala98               1                4.580000                  4.58
Zhisrisu83                  2                3.945000                  7.89
Zontibe81                   3                2.676667                  8.03

[576 rows x 3 columns]

In [51]:
#I'll sort these by the top Total Purchase and format the values. When I sort the values, I'll also specifically only include the top 5 values
top_purchase_summary_sort = top_purchase_summary.nlargest(5,"Total Purchase Value")
top_purchase_summary_sort["Average Purchase Price"] = top_purchase_summary_sort["Average Purchase Price"].astype(float).map("${:,.2f}".format)
top_purchase_summary_sort["Total Purchase Value"] = top_purchase_summary_sort["Total Purchase Value"].astype(float).map("${:,.2f}".format)
top_purchase_summary_sort

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

 ## Most Popular Items

### Calculations

For this step, I'll take the following actions

* Retrieve the Item ID, Item Name, and Item Price columns and place into a separate dataframe

* Use this dataframe to Group by Item ID and Item Name. Perform calculations to obtain 
    * purchase count
    * item price
    * total purchase value

In [52]:
#Take a look at my main dataframe
purchases_df

Purchase ID             SN  Age  Gender  Item ID  \
0              0        Lisim78   20    Male      108   
1              1    Lisovynya38   40    Male      143   
2              2     Ithergue48   24    Male       92   
3              3  Chamassasya86   24    Male      100   
4              4      Iskosia90   23    Male      131   
..           ...            ...  ...     ...      ...   
775          775     Aethedru70   21  Female       60   
776          776         Iral74   21    Male      164   
777          777     Yathecal72   20    Male       67   
778          778        Sisur91    7    Male      101   
779          779      Ennrian78   24    Male       50   

                                     Item Name  Price Age Group  
0    Extraction, Quickblade Of Trembling Hands   3.53     20-24  
1                            Frenzied Scimitar   1.56       40+  
2                                 Final Critic   4.88     20-24  
3                                  Blindscythe   3.27     20-24  
4                                         Fury   1.44     20-24  
..                                         ...    ...       ...  
775                                       Wolf   3.54     20-24  
776                           Exiled Doomblade   1.63     20-24  
777      Celeste, Incarnation of the Corrupted   3.46     20-24  
778                               Final Critic   4.19       <10  
779                                       Dawn   4.60     20-24  

[780 rows x 8 columns]

In [53]:
#Create a new dataframe with just a few columns: Item ID, Item Name, and Item Price columns
items_df=purchases_df.loc[:,["Item ID", "Item Name", "Price"]]
items_df

Item ID                                  Item Name  Price
0        108  Extraction, Quickblade Of Trembling Hands   3.53
1        143                          Frenzied Scimitar   1.56
2         92                               Final Critic   4.88
3        100                                Blindscythe   3.27
4        131                                       Fury   1.44
..       ...                                        ...    ...
775       60                                       Wolf   3.54
776      164                           Exiled Doomblade   1.63
777       67      Celeste, Incarnation of the Corrupted   3.46
778      101                               Final Critic   4.19
779       50                                       Dawn   4.60

[780 rows x 3 columns]

In [54]:
#groupby item id and item name
item_grpby = items_df.groupby(["Item ID", "Item Name"])
item_grpby.count()

Price
Item ID Item Name                             
0       Splinter                             4
1       Crucifer                             3
2       Verdict                              6
3       Phantomlight                         6
4       Bloodlord's Fetish                   5
...                                        ...
179     Wolf, Promise of the Moonwalker      6
180     Stormcaller                          1
181     Reaper's Toll                        5
182     Toothpick                            3
183     Dragon's Greatsword                  3

[183 rows x 1 columns]

In [55]:
#obtain the count of items per item ID/Item name combination
item_count = item_grpby["Price"].count()
item_count

Item ID  Item Name                      
0        Splinter                           4
1        Crucifer                           3
2        Verdict                            6
3        Phantomlight                       6
4        Bloodlord's Fetish                 5
                                           ..
179      Wolf, Promise of the Moonwalker    6
180      Stormcaller                        1
181      Reaper's Toll                      5
182      Toothpick                          3
183      Dragon's Greatsword                3
Name: Price, Length: 183, dtype: int64

In [56]:
#obtain the item info for the item ID/Item name combination. I'm using max because there should only be one item for each item ID/Item name combination and I simply want to return that value.
item_value = item_grpby["Price"].max()
item_value

Item ID  Item Name                      
0        Splinter                           1.28
1        Crucifer                           3.26
2        Verdict                            2.48
3        Phantomlight                       2.49
4        Bloodlord's Fetish                 1.70
                                            ... 
179      Wolf, Promise of the Moonwalker    4.48
180      Stormcaller                        3.36
181      Reaper's Toll                      1.66
182      Toothpick                          4.03
183      Dragon's Greatsword                1.09
Name: Price, Length: 183, dtype: float64

In [57]:
#obtain the sum total price for each item ID/Item name combination
item_total = item_grpby["Price"].sum()
item_total

Item ID  Item Name                      
0        Splinter                            5.12
1        Crucifer                            9.78
2        Verdict                            14.88
3        Phantomlight                       14.94
4        Bloodlord's Fetish                  8.50
                                            ...  
179      Wolf, Promise of the Moonwalker    26.88
180      Stormcaller                         3.36
181      Reaper's Toll                       8.30
182      Toothpick                          12.09
183      Dragon's Greatsword                 3.27
Name: Price, Length: 183, dtype: float64

### Summary

In this section, I'll take my calculations and place it a dataframe, format it and present only the top 5 items, based on total purchase count

In [58]:
#I'll put these values in a dataframe for presentation
popular_items_summary = pd.DataFrame({"Purchase Count": item_count,
                                 "Item Price": item_value,
                                 "Total Purchase Value": item_total})
popular_items_summary

Purchase Count  Item Price  \
Item ID Item Name                                                     
0       Splinter                                      4        1.28   
1       Crucifer                                      3        3.26   
2       Verdict                                       6        2.48   
3       Phantomlight                                  6        2.49   
4       Bloodlord's Fetish                            5        1.70   
...                                                 ...         ...   
179     Wolf, Promise of the Moonwalker               6        4.48   
180     Stormcaller                                   1        3.36   
181     Reaper's Toll                                 5        1.66   
182     Toothpick                                     3        4.03   
183     Dragon's Greatsword                           3        1.09   

                                         Total Purchase Value  
Item ID Item Name                                              
0       Splinter                                         5.12  
1       Crucifer                                         9.78  
2       Verdict                                         14.88  
3       Phantomlight                                    14.94  
4       Bloodlord's Fetish                               8.50  
...                                                       ...  
179     Wolf, Promise of the Moonwalker                 26.88  
180     Stormcaller                                      3.36  
181     Reaper's Toll                                    8.30  
182     Toothpick                                       12.09  
183     Dragon's Greatsword                              3.27  

[183 rows x 3 columns]

In [59]:
#I'll sort these by the top Total Purchase and format the values. When I sort the values, I'll also specifically only include the top 5 values
popular_items_summary_sort = popular_items_summary.nlargest(5,"Purchase Count")
popular_items_summary_sort["Item Price"] = popular_items_summary_sort["Item Price"].astype(float).map("${:,.2f}".format)
popular_items_summary_sort["Total Purchase Value"] = popular_items_summary_sort["Total Purchase Value"].astype(float).map("${:,.2f}".format)
popular_items_summary_sort

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
108     Extraction, Quickblade Of Trembling Hands                  9   
145     Fiery Glass Crusader                                       9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
145     Fiery Glass Crusader                              $4.58   
19      Pursuit, Cudgel of Necromancy                     $1.02   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
145     Fiery Glass Crusader                                       $41.22  
19      Pursuit, Cudgel of Necromancy                               $8.16

## Most Profitable Items

### Calculations

There were no calculations to make in this section.  All of the calculations were completed in section 8.

### Summary

For this section, I'm simply taking my previous data frame from section 8, and re-sorting it to sort by total purchase value in descending order.

In [60]:
#I'll sort these by the top Total Purchase and format the values. When I sort the values, I'll also specifically only include the top 5 values
popular_items_summary_sort = popular_items_summary.nlargest(5,"Total Purchase Value")
popular_items_summary_sort["Item Price"] = popular_items_summary_sort["Item Price"].astype(float).map("${:,.2f}".format)
popular_items_summary_sort["Total Purchase Value"] = popular_items_summary_sort["Total Purchase Value"].astype(float).map("${:,.2f}".format)
popular_items_summary_sort

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

## Analysis Conclusion

After performing all of this analysis, there are a few takeaways that can be observed from the data.

**1. Males buy more items, but Females or Non-Disclosed users spend more on average**

In looking at the summary from 4.2, I can see that males dwarf all other gender categories when it comes to total purchase values, simply because there are more male users and they are buying more items.  However, females and non-disclosed users genearlly spend MORE on average when they do make purchases. Females have an avergage purchase price of \\$3.20 and average total purchase of \\$4.47.  Non-disclosed users have an avergage purchase price of \\$3.35 and average total purchase of \\$4.56.  Males have an avergage purchase price of \\$3.02" and average total purchase of \\$4.07.  What this tells us is that if there was more of a focus on getting non-male users to buy more items, then more money could be made.

**2. The age 20-24 demographic is an important one to focus on**

In looking at the summaries from 5.2 and 6.2, we can see that users that are betweent the ages of 20-24 make up the largest group of users, at nearly 45%. Almost half of the users come from a single age group, so it would be wise to focus advertising and other methods of engagement on that group.  The next important group would be all of the groups under the age of 20.  In total, these groups comprise over 25% of the users.  More advertising to kids or young adults will help to maintain this group's engagement. This information is also supported in the summary in 6.2 where dollar figures are attached to the age demographics, and they support the above assertions, when looking at total purchase value.

**3. Oathbreaker, Last Hope of the Breaking Storm and Nirvana are popular when it comes to popularity and profitability**

In looking at the summaries from 8.2 and 9.2,  both of these items were on the top when it comes to popularity and profitability.  People are buying these items frequently and thier current price point is helping the bottom line.  It would be good to highlight these games to encourage more users to purchase it and to help earn more money.